File: platform/android/lynx_android/src/main/java/com/lynx/tasm/behavior/ui/frame/FrameShadowNode.java
// Copyright 2026 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

package com.lynx.tasm.behavior.ui.frame;

import com.lynx.tasm.base.TraceEvent;
import com.lynx.tasm.base.trace.TraceEventDef;
import com.lynx.tasm.behavior.LynxGeneratorName;
import com.lynx.tasm.behavior.LynxShadowNode;
import com.lynx.tasm.behavior.shadow.AlignContext;
import com.lynx.tasm.behavior.shadow.AlignParam;
import com.lynx.tasm.behavior.shadow.CustomLayoutShadowNode;
import com.lynx.tasm.behavior.shadow.CustomMeasureFunc;
import com.lynx.tasm.behavior.shadow.MeasureContext;
import com.lynx.tasm.behavior.shadow.MeasureMode;
import com.lynx.tasm.behavior.shadow.MeasureParam;
import com.lynx.tasm.behavior.shadow.MeasureResult;
import java.util.HashMap;

@LynxGeneratorName(packageName = "com.lynx.tasm.behavior.ui.frame")
@LynxShadowNode(tagName = "frame")
public class FrameShadowNode extends CustomLayoutShadowNode implements CustomMeasureFunc {
  private static final String TAG = "FrameShadowNode";
  private int mIntrinsicWidth;
  private int mIntrinsicHeight;

  @Override
  public void attachNativePtr(long ptr) {
    setCustomMeasureFunc(this);
    super.attachNativePtr(ptr);
  }

  /**
   * @brief Update the intrinsic content size of the FrameShadowNode.
   * @param width The intrinsic width of the FrameShadowNode.
   * @param height The intrinsic height of the FrameShadowNode.
   */
  public void updateIntrinsicContentSize(int width, int height) {
    if (TraceEvent.isTracingStarted()) {
      HashMap<String, String> traceProps = new HashMap<>();
      traceProps.put("sign", String.valueOf(getSignature()));
      traceProps.put("old_width", String.valueOf(mIntrinsicWidth));
      traceProps.put("old_height", String.valueOf(mIntrinsicHeight));
      traceProps.put("new_width", String.valueOf(width));
      traceProps.put("new_height", String.valueOf(height));
      TraceEvent.instant(TraceEvent.CATEGORY_DEFAULT,
          "FrameIntrinsicTrace.FrameShadowNode.updateIntrinsicContentSize", traceProps);
    }
    if (width != mIntrinsicWidth || height != mIntrinsicHeight) {
      mIntrinsicWidth = width;
      mIntrinsicHeight = height;
      markDirty();
    }
  }

  @Override
  public MeasureResult measure(MeasureParam param, MeasureContext context) {
    if (TraceEvent.isTracingStarted()) {
      HashMap<String, String> args = new HashMap<>();
      args.put("input_width", String.valueOf(param.mWidth));
      args.put("input_height", String.valueOf(param.mHeight));
      args.put("input_width_mode", param.mWidthMode.name());
      args.put("input_height_mode", param.mHeightMode.name());
      TraceEvent.beginSection(TraceEventDef.FRAME_SHADOW_NODE_MEASURE, args);
    }
    MeasureResult result;
    if (mIntrinsicWidth == 0 && mIntrinsicHeight == 0) {
      result = new MeasureResult(param.mWidth, param.mHeight);
    } else {
      float width = param.mWidthMode == MeasureMode.EXACTLY ? param.mWidth : mIntrinsicWidth;
      float height = param.mHeightMode == MeasureMode.EXACTLY ? param.mHeight : mIntrinsicHeight;
      result = new MeasureResult(width, height);
    }
    if (TraceEvent.isTracingStarted()) {
      HashMap<String, String> args = new HashMap<>();
      args.put("width", String.valueOf(result.getWidthResult()));
      args.put("height", String.valueOf(result.getHeightResult()));
      TraceEvent.endSection(
          TraceEvent.CATEGORY_DEFAULT, TraceEventDef.FRAME_SHADOW_NODE_MEASURE, args);
    }
    return result;
  }

  @Override
  public void align(AlignParam param, AlignContext context) {}
}

File: platform/harmony/lynx_harmony/src/main/cpp/shadow_node/frame_shadow_node.cc
// Copyright 2026 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

#include "platform/harmony/lynx_harmony/src/main/cpp/shadow_node/frame_shadow_node.h"

#include "base/include/float_comparison.h"
#include "base/trace/native/trace_event.h"
#include "core/base/harmony/harmony_trace_event_def.h"
#include "core/base/lynx_trace_categories.h"

namespace lynx {
namespace tasm {
namespace harmony {

FrameShadowNode::FrameShadowNode(int sign, const std::string& tag)
    : ShadowNode(sign, tag) {
  SetCustomMeasureFunc(this);
}

LayoutResult FrameShadowNode::Measure(float width, MeasureMode width_mode,
                                      float height, MeasureMode height_mode,
                                      bool final_measure) {
  TRACE_EVENT_BEGIN(LYNX_TRACE_CATEGORY, FRAME_SHADOW_NODE_MEASURE, "sign",
                    Signature(), "input_width", width, "input_height", height,
                    "input_width_mode", width_mode, "input_height_mode",
                    height_mode);

  LayoutResult result(width, height, 0.f);
  if (!base::IsZero(intrinsic_width_) || !base::IsZero(intrinsic_height_)) {
    result.width_ =
        width_mode == MeasureMode::Definite ? width : intrinsic_width_;
    result.height_ =
        height_mode == MeasureMode::Definite ? height : intrinsic_height_;
  }

  TRACE_EVENT_END(LYNX_TRACE_CATEGORY, [&result](perfetto::EventContext ctx) {
    ctx.event()->add_debug_annotations("width", std::to_string(result.width_));
    ctx.event()->add_debug_annotations("height",
                                       std::to_string(result.height_));
  });
  return result;
}

void FrameShadowNode::UpdateIntrinsicSize(float intrinsic_width,
                                          float intrinsic_height) {
  TRACE_EVENT_INSTANT(LYNX_TRACE_CATEGORY,
                      FRAME_SHADOW_NODE_UPDATE_INTRINSIC_SIZE, "sign",
                      Signature(), "old_width", intrinsic_width_, "old_height",
                      intrinsic_height_, "new_width", intrinsic_width,
                      "new_height", intrinsic_height);

  if (base::FloatsEqual(intrinsic_width_, intrinsic_width) &&
      base::FloatsEqual(intrinsic_height_, intrinsic_height)) {
    return;
  }
  intrinsic_width_ = intrinsic_width;
  intrinsic_height_ = intrinsic_height;
  MarkDirty();
}

}  // namespace harmony
}  // namespace tasm
}  // namespace lynx
